# Association rule mining basics

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv("Online Retail.csv")
data.head(20)

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5     536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6     536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7     536366     22633               HAND WARMER UNION JACK         6   
8     536366     22632            HAND WARMER RED POLKA DOT         6   
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10    536367     22745           POPPY'S PLAYHOUSE BEDROOM          6   
11    536367     22748            POPPY'S PLAYHOUSE KITCHEN         6   
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13    536367     22310              IVORY KNITTED MUG COSY          6   
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
15    536367     22623        BOX OF VINTAGE JIGSAW BLOCKS          3   
16    536367     22622       BOX OF VINTAGE ALPHABET BLOCKS         2   
17    536367     21754             HOME BUILDING BLOCK WORD         3   
18    536367     21755             LOVE BUILDING BLOCK WORD         3   
19    536367     21777          RECIPE BOX WITH METAL HEART         4   

         InvoiceDate  UnitPrice  CustomerID         Country  
0   01.12.2010 08:26       2.55     17850.0  United Kingdom  
1   01.12.2010 08:26       3.39     17850.0  United Kingdom  
2   01.12.2010 08:26       2.75     17850.0  United Kingdom  
3   01.12.2010 08:26       3.39     17850.0  United Kingdom  
4   01.12.2010 08:26       3.39     17850.0  United Kingdom  
5   01.12.2010 08:26       7.65     17850.0  United Kingdom  
6   01.12.2010 08:26       4.25     17850.0  United Kingdom  
7   01.12.2010 08:28       1.85     17850.0  United Kingdom  
8   01.12.2010 08:28       1.85     17850.0  United Kingdom  
9   01.12.2010 08:34       1.69     13047.0  United Kingdom  
10  01.12.2010 08:34       2.10     13047.0  United Kingdom  
11  01.12.2010 08:34       2.10     13047.0  United Kingdom  
12  01.12.2010 08:34       3.75     13047.0  United Kingdom  
13  01.12.2010 08:34       1.65     13047.0  United Kingdom  
14  01.12.2010 08:34       4.25     13047.0  United Kingdom  
15  01.12.2010 08:34       4.95     13047.0  United Kingdom  
16  01.12.2010 08:34       9.95     13047.0  United Kingdom  
17  01.12.2010 08:34       5.95     13047.0  United Kingdom  
18  01.12.2010 08:34       5.95     13047.0  United Kingdom  
19  01.12.2010 08:34       7.95     13047.0  United Kingdom

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


**Q:Какие строки и стоит удалить и как это сделать?**

In [5]:
data.dropna(subset=['Description'],inplace=True)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 540455 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    540455 non-null  object 
 1   StockCode    540455 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     540455 non-null  int64  
 4   InvoiceDate  540455 non-null  object 
 5   UnitPrice    540455 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      540455 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 37.1+ MB


**Q: Как посчитать статистику распределения количества предметов по чекам (invoice)?**

In [7]:
data.groupby(['InvoiceNo']).size().describe()

count    24446.000000
mean        22.108116
std         44.009335
min          1.000000
25%          3.000000
50%         11.000000
75%         25.000000
max       1114.000000
dtype: float64

### Data preprocessing

In [8]:
#'Country', 'InvoiceDate','CustomerID'

data_by_invoice = data.groupby(['InvoiceNo','Country'])['Description'].apply(lambda x: ','.join(x)).reset_index()
data_by_invoice

InvoiceNo         Country  \
0        536365  United Kingdom   
1        536366  United Kingdom   
2        536367  United Kingdom   
3        536368  United Kingdom   
4        536369  United Kingdom   
...         ...             ...   
24441   C581484  United Kingdom   
24442   C581490  United Kingdom   
24443   C581499  United Kingdom   
24444   C581568  United Kingdom   
24445   C581569  United Kingdom   

                                             Description  
0      WHITE HANGING HEART T-LIGHT HOLDER,WHITE METAL...  
1       HAND WARMER UNION JACK,HAND WARMER RED POLKA DOT  
2      ASSORTED COLOUR BIRD ORNAMENT,POPPY'S PLAYHOUS...  
3      JAM MAKING SET WITH JARS,RED COAT RACK PARIS F...  
4                               BATH BUILDING BLOCK WORD  
...                                                  ...  
24441                        PAPER CRAFT . LITTLE BIRDIE  
24442  VICTORIAN GLASS HANGING T-LIGHT,ZINC T-LIGHT H...  
24443                                             Manual  
24444                         VICTORIAN SEWING BOX LARGE  
24445  HANGING HEART JAR T-LIGHT HOLDER,36 PENCILS TU...  

[24446 rows x 3 columns]

In [9]:
data_by_invoice['Country'].value_counts()

United Kingdom          22040
Germany                   603
France                    461
EIRE                      360
Belgium                   119
Spain                     105
Netherlands               101
Switzerland                74
Portugal                   71
Australia                  69
Italy                      55
Finland                    48
Sweden                     46
Norway                     40
Channel Islands            33
Japan                      28
Poland                     24
Denmark                    21
Cyprus                     20
Austria                    19
Hong Kong                  15
Unspecified                13
Singapore                  10
Malta                      10
Israel                      9
USA                         7
Iceland                     7
Greece                      6
Canada                      6
European Community          5
Czech Republic              5
Lithuania                   4
Bahrain                     4
United Ara

#### Get country data

In [10]:
data_by_invoice_france = data_by_invoice[data_by_invoice.Country == 'France']
data_by_invoice_france.head()

InvoiceNo Country                                        Description
5      536370  France  ALARM CLOCK BAKELIKE PINK,ALARM CLOCK BAKELIKE...
273    536852  France  PICTURE DOMINOES,MINI JIGSAW SPACEBOY,MINI JIG...
311    536974  France  EDWARDIAN PARASOL BLACK,EDWARDIAN PARASOL PINK...
360    537065  France  HOT WATER BOTTLE BABUSHKA ,BREAD BIN DINER STY...
543    537463  France  JAM MAKING SET PRINTED,SET/4 SKULL BADGES,ROUN...

#### Dummy encoding <br>
<img src="dummy.png" width="500">

In [11]:
dummy_data_by_invoice_france = data_by_invoice_france['Description'].str.get_dummies(',')
data_france = pd.concat([data_by_invoice_france, dummy_data_by_invoice_france], axis = 1)
data_france.head(5)

InvoiceNo Country                                        Description  \
5      536370  France  ALARM CLOCK BAKELIKE PINK,ALARM CLOCK BAKELIKE...   
273    536852  France  PICTURE DOMINOES,MINI JIGSAW SPACEBOY,MINI JIG...   
311    536974  France  EDWARDIAN PARASOL BLACK,EDWARDIAN PARASOL PINK...   
360    537065  France  HOT WATER BOTTLE BABUSHKA ,BREAD BIN DINER STY...   
543    537463  France  JAM MAKING SET PRINTED,SET/4 SKULL BADGES,ROUN...   

      50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
5                                 0                   0   
273                               0                   0   
311                               0                   0   
360                               0                   0   
543                               0                   0   

      I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
5                               0                         0   
273                             0                         0   
311                             0                         0   
360                             0                         0   
543                             0                         0   

      SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
5                                   1                        0   
273                                 0                        0   
311                                 0                        0   
360                                 0                        0   
543                                 0                        0   

      TRELLIS COAT RACK  ...  WRAP VINTAGE PETALS  DESIGN  \
5                     0  ...                            0   
273                   0  ...                            0   
311                   0  ...                            0   
360                   0  ...                            0   
543                   0  ...                            0   

     YELLOW COAT RACK PARIS FASHION  YELLOW GIANT GARDEN THERMOMETER  \
5                                 0                                0   
273                               0                                0   
311                               0                                0   
360                               0                                0   
543                               0                                0   

     YELLOW SHARK HELICOPTER  ZINC  STAR T-LIGHT HOLDER   \
5                          0                           0   
273                        0                           0   
311                        0                           0   
360                        0                           0   
543                        0                           0   

     ZINC FOLKART SLEIGH BELLS  ZINC HERB GARDEN CONTAINER  \
5                            0                           0   
273                          0                           0   
311                          0                           0   
360                          0                           0   
543                          0                           0   

     ZINC METAL HEART DECORATION  ZINC T-LIGHT HOLDER STAR LARGE  \
5                              0                               0   
273                            0                               0   
311                            0                               0   
360                            0                               0   
543                            0                               0   

     ZINC T-LIGHT HOLDER STARS SMALL  
5                                  0  
273                                0  
311                                0  
360                                0  
543                                0  

[5 rows x 1568 columns]

In [12]:
data_france.columns

Index(['InvoiceNo', 'Country', 'Description', ' 50'S CHRISTMAS GIFT BAG LARGE',
       ' DOLLY GIRL BEAKER', ' I LOVE LONDON MINI BACKPACK',
       ' NINE DRAWER OFFICE TIDY', ' SET 2 TEA TOWELS I LOVE LONDON ',
       ' SPACEBOY BABY GIFT SET', ' TRELLIS COAT RACK',
       ...
       'WRAP VINTAGE PETALS  DESIGN', 'YELLOW COAT RACK PARIS FASHION',
       'YELLOW GIANT GARDEN THERMOMETER', 'YELLOW SHARK HELICOPTER',
       'ZINC  STAR T-LIGHT HOLDER ', 'ZINC FOLKART SLEIGH BELLS',
       'ZINC HERB GARDEN CONTAINER', 'ZINC METAL HEART DECORATION',
       'ZINC T-LIGHT HOLDER STAR LARGE', 'ZINC T-LIGHT HOLDER STARS SMALL'],
      dtype='object', length=1568)

In [13]:
#Еще один способ сделать то же самое

data[data.Country == 'France'].pivot_table(index='InvoiceNo', columns='Description', aggfunc=any, fill_value=False).astype(int)

Country                     \
Description  50'S CHRISTMAS GIFT BAG LARGE  DOLLY GIRL BEAKER   
InvoiceNo                                                       
536370                                   0                  0   
536852                                   0                  0   
536974                                   0                  0   
537065                                   0                  0   
537463                                   0                  0   
...                                    ...                ...   
C579532                                  0                  0   
C579562                                  0                  0   
C580161                                  0                  0   
C580263                                  0                  0   
C581316                                  0                  0   

                                                                   \
Description  I LOVE LONDON MINI BACKPACK  NINE DRAWER OFFICE TIDY   
InvoiceNo                                                           
536370                                 0                        0   
536852                                 0                        0   
536974                                 0                        0   
537065                                 0                        0   
537463                                 0                        0   
...                                  ...                      ...   
C579532                                0                        0   
C579562                                0                        0   
C580161                                0                        0   
C580263                                0                        0   
C581316                                0                        0   

                                                                      \
Description  SET 2 TEA TOWELS I LOVE LONDON   SPACEBOY BABY GIFT SET   
InvoiceNo                                                              
536370                                     1                       0   
536852                                     0                       0   
536974                                     0                       0   
537065                                     0                       0   
537463                                     0                       0   
...                                      ...                     ...   
C579532                                    0                       0   
C579562                                    0                       0   
C580161                                    0                       0   
C580263                                    0                       0   
C581316                                    0                       0   

                                                       \
Description  TRELLIS COAT RACK 10 COLOUR SPACEBOY PEN   
InvoiceNo                                               
536370                       0                      0   
536852                       0                      0   
536974                       0                      0   
537065                       0                      0   
537463                       0                      0   
...                        ...                    ...   
C579532                      0                      0   
C579562                      0                      0   
C580161                      0                      0   
C580263                      0                      0   
C581316                      0                      0   

                                                                  ...  \
Description 12 COLOURED PARTY BALLOONS 12 EGG HOUSE PAINTED WOOD  ...   
InvoiceNo                                                         ...   
536370                               0                         0  ...   
536852                               0                   

### Extract rules with Mlxtend (machine learning extensions) 

Python library of useful tools for the day-to-day data science tasks.

http://rasbt.github.io/mlxtend/api_subpackages/mlxtend.frequent_patterns/

### Find frequent itemsets and rules

In [14]:
from mlxtend.frequent_patterns import apriori, association_rules

frequent_itemsets = apriori(data_france.drop(columns=['InvoiceNo','Country','Description']), 
                            min_support=0.05, use_colnames=True)

frequent_itemsets

support                                           itemsets
0    0.060738                      (4 TRADITIONAL SPINNING TOPS)
1    0.084599                       (ALARM CLOCK BAKELIKE GREEN)
2    0.086768                        (ALARM CLOCK BAKELIKE PINK)
3    0.080260                        (ALARM CLOCK BAKELIKE RED )
4    0.058568                       (ASSORTED COLOUR MINI CASES)
..        ...                                                ...
125  0.071584  (SET/20 RED RETROSPOT PAPER NAPKINS , POSTAGE,...
126  0.071584  (SET/20 RED RETROSPOT PAPER NAPKINS , POSTAGE,...
127  0.086768  (SET/6 RED SPOTTY PAPER PLATES, POSTAGE, SET/6...
128  0.084599  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...
129  0.069414  (SET/20 RED RETROSPOT PAPER NAPKINS , POSTAGE,...

[130 rows x 2 columns]

In [15]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)
rules.head()

antecedents                   consequents  \
0  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE PINK)   
1   (ALARM CLOCK BAKELIKE PINK)  (ALARM CLOCK BAKELIKE GREEN)   
2  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE RED )   
3   (ALARM CLOCK BAKELIKE RED )  (ALARM CLOCK BAKELIKE GREEN)   
4  (ALARM CLOCK BAKELIKE GREEN)                     (POSTAGE)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.084599            0.086768  0.062907    0.743590  8.569872   
1            0.086768            0.084599  0.062907    0.725000  8.569872   
2            0.084599            0.080260  0.067245    0.794872  9.903673   
3            0.080260            0.084599  0.067245    0.837838  9.903673   
4            0.084599            0.674620  0.071584    0.846154  1.254267   

   leverage  conviction  
0  0.055566    3.561605  
1  0.055566    3.328732  
2  0.060455    4.483731  
3  0.060455    5.644975  
4  0.014512    2.114967

### Add antecedent lenght column

In [16]:
rules["antecedent_len"] = rules["antecedents"].apply(lambda x: len(x))
rules.head()

antecedents                   consequents  \
0  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE PINK)   
1   (ALARM CLOCK BAKELIKE PINK)  (ALARM CLOCK BAKELIKE GREEN)   
2  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE RED )   
3   (ALARM CLOCK BAKELIKE RED )  (ALARM CLOCK BAKELIKE GREEN)   
4  (ALARM CLOCK BAKELIKE GREEN)                     (POSTAGE)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.084599            0.086768  0.062907    0.743590  8.569872   
1            0.086768            0.084599  0.062907    0.725000  8.569872   
2            0.084599            0.080260  0.067245    0.794872  9.903673   
3            0.080260            0.084599  0.067245    0.837838  9.903673   
4            0.084599            0.674620  0.071584    0.846154  1.254267   

   leverage  conviction  antecedent_len  
0  0.055566    3.561605               1  
1  0.055566    3.328732               1  
2  0.060455    4.483731               1  
3  0.060455    5.644975               1  
4  0.014512    2.114967               1

### Filter rules

**Q: Выведите правила, у которых метрика lift > 3го квантиля? Сколько их?**

In [18]:
rules['lift'].describe()

count    141.000000
mean       5.185359
std        3.436225
min        1.074678
25%        1.297026
50%        5.122222
75%        8.195556
max       14.719029
Name: lift, dtype: float64

In [ ]:
rules_lift_75 = rules[rules['lift'] > 8.195] #8.195 - это значение третьего квантиля
print (len(rules_lift_75))
rules_lift_75.head(10)

**Q: Выведите правила, у которых метрика lift > 1.297 (1 квантиль), поддержка > 0.07, а посылка правила включает *только*  'PLASTERS IN TIN WOODLAND ANIMALS'. Сколько их?**

In [22]:
first_quantile = rules.quantile(q = 0.25)['lift']
rules[(rules['lift'] > first_quantile) &
      (rules['support'] > 0.07) &
      (rules['consequents'] == {'PLASTERS IN TIN WOODLAND ANIMALS'})]

antecedents  \
37            (PLASTERS IN TIN CIRCUS PARADE )   
40                  (PLASTERS IN TIN SPACEBOY)   
102  (POSTAGE, PLASTERS IN TIN CIRCUS PARADE )   
105        (POSTAGE, PLASTERS IN TIN SPACEBOY)   

                            consequents  antecedent support  \
37   (PLASTERS IN TIN WOODLAND ANIMALS)            0.147505   
40   (PLASTERS IN TIN WOODLAND ANIMALS)            0.119306   
102  (PLASTERS IN TIN WOODLAND ANIMALS)            0.125813   
105  (PLASTERS IN TIN WOODLAND ANIMALS)            0.097614   

     consequent support   support  confidence      lift  leverage  conviction  \
37             0.145336  0.086768    0.588235  4.047410  0.065330    2.075612   
40             0.145336  0.088937    0.745455  5.129172  0.071598    3.357608   
102            0.145336  0.071584    0.568966  3.914822  0.053298    1.982820   
105            0.145336  0.071584    0.733333  5.045771  0.057397    3.204989   

     antecedent_len  
37                1  
40                1  
102               2  
105               2

**Q\*: Выведите правила, у которых метрика confidence > 0.7, поддержка > 0.05, а следствие правила включает 'POSTAGE'. Сколько их?**

In [23]:
rules[(rules['confidence'] > 0.7) &
      (rules['support'] > 0.05) &
      (rules['consequents'].astype(str).str.contains('POSTAGE'))]

antecedents  \
4                         (ALARM CLOCK BAKELIKE GREEN)   
7                          (ALARM CLOCK BAKELIKE PINK)   
8                          (ALARM CLOCK BAKELIKE RED )   
9                         (ASSORTED COLOUR MINI CASES)   
12                          (COFFEE MUG APPLES DESIGN)   
13                              (DOLLY GIRL LUNCH BOX)   
16                           (JUMBO BAG RED RETROSPOT)   
17                        (JUMBO BAG WOODLAND ANIMALS)   
21                            (LUNCH BAG APPLE DESIGN)   
23                           (LUNCH BAG RED RETROSPOT)   
26                        (LUNCH BAG SPACEBOY DESIGN )   
27                                (LUNCH BAG WOODLAND)   
28                 (LUNCH BOX WITH CUTLERY RETROSPOT )   
30                           (MINI PAINT SET VINTAGE )   
31                        (PACK OF 6 SKULL PAPER CUPS)   
32                   (PACK OF 72 RETROSPOT CAKE CASES)   
33                           (PAPER BUNTING RETROSPOT)   
38                    (PLASTERS IN TIN CIRCUS PARADE )   
41                          (PLASTERS IN TIN SPACEBOY)   
42                         (PLASTERS IN TIN STRONGMAN)   
43                  (PLASTERS IN TIN WOODLAND ANIMALS)   
44                                (RABBIT NIGHT LIGHT)   
45                       (RED RETROSPOT CHARLOTTE BAG)   
46                          (RED RETROSPOT MINI CASES)   
47                          (RED RETROSPOT PICNIC BAG)   
48                     (RED TOADSTOOL LED NIGHT LIGHT)   
49                          (REGENCY CAKESTAND 3 TIER)   
50                (ROUND SNACK BOXES SET OF 4 FRUITS )   
51               (ROUND SNACK BOXES SET OF4 WOODLAND )   
52               (SET OF 2 TEA TOWELS APPLE AND PEARS)   
53                     (SET OF 9 BLACK SKULL BALLOONS)   
54                 (SET/10 RED POLKADOT PARTY CANDLES)   
55               (SET/20 RED RETROSPOT PAPER NAPKINS )   
56                       (SET/6 RED SPOTTY PAPER CUPS)   
57                     (SET/6 RED SPOTTY PAPER PLATES)   
58                               (SPACEBOY LUNCH BOX )   
59                 (STRAWBERRY LUNCH BOX WITH CUTLERY)   
60                           (TEA PARTY BIRTHDAY CARD)   
61                            (WOODLAND CHARLOTTE BAG)   
75   (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
81   (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
83                        (ALARM CLOCK BAKELIKE GREEN)   
84                         (ALARM CLOCK BAKELIKE RED )   
86   (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...   
89                         (ALARM CLOCK BAKELIKE RED )   
90   (LUNCH BAG APPLE DESIGN, LUNCH BAG RED RETROSPOT)   
97   (PLASTERS IN TIN SPACEBOY, PLASTERS IN TIN CIR...   
101  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...   
103  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...   
108  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
113  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
118  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...   
120                    (SET/6 RED SPOTTY PAPER PLATES)   
121                      (SET/6 RED SPOTTY PAPER CUPS)   
130  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
133  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
134  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   

                                  consequents  antecedent support  \
4                                   (POSTAGE)            0.084599   
7                                   (POSTAGE)            0.086768   
8                                   (POSTAGE)            0.080260   
9                                   (POSTAGE)            0.058568   
12                                  (POSTAGE)            0.060738   
13                                  (POSTAGE)            0.086768   
16                                  (POSTAGE)            0.082430   
17                                  (POSTAGE)            0.065076   
21                                  (POSTAGE)            